src/lib.rs

```diff
@@ -336,28 +334,15 @@ impl<Types: SwarmTypes> Future for IpfsFuture<Types> {
             // libp2p uses std::future ... I couldn't figure out way to wrap it as compat,
             // box it and fuse it to for it to be used with futures::select!
 
-            {
-                let pin = Pin::new(&mut self.exit_events);
-
-                if let Poll::Ready(Some(IpfsEvent::Exit)) = pin.poll_next(ctx) {
-                    return Poll::Ready(());
-                }
+            if let Ok(IpfsEvent::Exit) = self.exit_events.try_recv() {
+                return Poll::Ready(());
             }
 
-            {
-                loop {
-                    let pin = Pin::new(&mut self.repo_events);
-                    match pin.poll_next(ctx) {
-                        Poll::Ready(Some(RepoEvent::WantBlock(cid))) => self.swarm.want_block(cid),
-                        Poll::Ready(Some(RepoEvent::ProvideBlock(cid))) => {
-                            self.swarm.provide_block(cid)
-                        }
-                        Poll::Ready(Some(RepoEvent::UnprovideBlock(cid))) => {
-                            self.swarm.stop_providing_block(&cid)
-                        }
-                        Poll::Ready(None) => panic!("other side closed the repo_events?"),
-                        Poll::Pending => break,
-                    }
+            while let Ok(event) = self.repo_events.try_recv() {
```

Comment:
Just noticed that crossbeam channels cannot be used in this context since it cannot awake this IpfsFuture task when swarm has no reason to wake up and something gets pushed into the channels.

Reply:
nice catch! thanks